working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
+declared type of a variable or parameter, as the return type of a function
+or getter, as a type argument in a type, or as the on-type of an extension.
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
+where `v` is a fresh variable whose declared type
+is the on-type corresponding to `Ext<S1, .. Sm>`, and similarly for
+instance getters and operators. This rule also applies when a member access
+implicitly has the receiver `this`, and the static type of `this` is an
+extension type (*which can only occur in an extension type member
+declaration*).
+
+*That is, when the type of an expression is an extension type, all method
+invocations on that expression will invoke an extension method declared by
+that extension, and similarly for other member accesses. In particular, we
+cannot invoke an instance member when the receiver type is an extension
+type (unless the the extension type enables them explicitly, cf. the
+show/hide part specified in a later section).*
+
+Let `E` be an extension declaration named `Ext` with type parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>k</sub> extends B<sub>k</sub></code>
+and on-type clause `on T`. Then we say that the _declared on-type_ of `Ext`
+is `T`, and the _instantiated on-type_ of 
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code> 
+is
+<code>[S<sub>1</sub>/X<sub>1</sub>, .. S<sub>k</sub>/X<sub>k</sub>]T</code>.
+We will omit 'declared' and 'instantiated' from the phrase when it is clear
+from the context whether we are talking about the extension itself or a
+particular instantiation of a generic extension. For non-generic
+extensions, the on-type is the same in either case.
+
+An extension type `E` of the form 
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is a subtype of `Object?`, and a proper supertype of the
```

Comment:
What about `Object`?  On the one hand, it feels like not exposing the nullability of the on type is good for abstraction, but on the other hand, it might be annoying for this entire class of types to be forbidden in places requiring `extends Object`.

Reply:
Allowing null values to flow to `Object` locations feels like a non-starter to me.  Not sure what to do about extensions.  You could make it explicit.  `extension E on T` means `E` is a subtype of `Object?` but `extension E extends Object on T` means `E` is a subtype of `Object` (and T must be a subtype of `Object`).  Kind of noisy though.

Reply:
We would also have to consider extension types which are potentially nullable:
```dart
extension type Ext<T> on T {
  ...
}
void foo<S>() {
  Ext<S> e = null;  // Ext<S> is neither nullable nor non-nullable, it's potentially nullable.
}
```
It's not new `FutureOr<T>` is also potentially nullable, so maybe it's not a complication.

It is a breach of abstraction.

Reply:
@leafpetersen wrote:
> Allowing null values to flow to `Object` locations feels like a non-starter to me.

Indeed. I should have been more explicit, but this was exactly the reason why I suggested that we could use `Object` as the supertype of an extension type _only_ in the case where the on-type is non-nullable.

I created https://github.com/dart-lang/language/issues/1490 in order to subcontract this discussion.